pkg/cluster/registry_cert.go

```diff
@@ -0,0 +1,129 @@
+package cluster
+
+import (
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	certsphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/renewal"
+)
+
+func (c *Controller) RotateRegistryCert() error {
+	ns := c.Config.RegistryCertNamespace
+	name := c.Config.RegistryCertSecret
+
+	if ns == "" || name == "" {
+		c.Log.Debugf("Registry namespace and secret not set, skipping certificate renewal")
+		return nil
+	}
+
+	cert, err := c.readRegistryCert(ns, name)
+	if err != nil {
+		return errors.Wrapf(err, "load existing certificate")
+	}
+	if cert == nil {
+		c.Log.Debugf("Registry cert not found")
```

Comment:
why would it not be found? should this be a higher level?

Reply:
If you don't have the registry enabled, which some people don't when running in online mode.